File: npm/src/report/jsonReport.js
// @ts-check
/**
 * 기계가 읽는 꼴. 평가 루프의 0층 입력이다. 파이썬 json.dumps(indent=2, ensure_ascii=False) 와 같은 글자다.
 * 지적마다 `exemplar` 를 붙인다. AI 는 규칙 이름과 이유만으로 고칠 때보다 본떠서 고칠 때 결과가 낫다.
 */
import { exemplarFor } from "../data/exemplars.js";
import { findingAsDict } from "../rules/finding.js";
import { exemplarInRegister } from "./registerMatch.js";

/** @param {import("../rules/finding.js").Finding} finding @param {string | null | undefined} register @param {string | null | undefined} preset */
function findingWithExemplar(finding, register, preset) {
  const data = findingAsDict(finding);
  const exemplar = exemplarFor(finding.rule, preset);
  if (exemplar) {
    const adapted = exemplarInRegister(exemplar, register);
    data.exemplar = { before: adapted.before, after: adapted.after, moved: adapted.moved };
  }
  return data;
}

/**
 * @param {Map<string, import("../rules/finding.js").Finding[]>} results
 * @param {string | null} [configLabel]
 * @param {Map<string, string> | null} [registers]
 * @param {string | null} [preset]
 */
export function renderJson(results, configLabel = null, registers = null, preset = null) {
  const files = [...results].map(([path, findings]) => ({
    path,
    findings: findings.map((finding) => findingWithExemplar(finding, registers?.get(path), preset)),
  }));
  /** @type {Record<string, unknown>} */
  const data = { version: 1 };
  if (configLabel !== null) data.config = configLabel;
  data.files = files;
  return JSON.stringify(data, null, 2);
}

File: src/hanlint/report/jsonReport.py
"""기계가 읽는 꼴. 평가 루프의 0층 입력이다. textlint 의 message 와 필드가 대응한다.

지적마다 `exemplar` 를 붙인다. 그 규칙의 고치기 전과 후의 짝이다. 사람이 읽는 꼴은 규칙마다 한 줄로
접지만 기계는 되풀이가 싸다. 본보기 유무의 실제 수정 차이는 tests/_attempts/exemplarLift 에서 따로 잰다.
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from ..audit import AuditResult
from ..data import Exemplar, exemplarFor
from ..rules import Finding
from .registerMatch import exemplarInRegister


def findingWithExemplar(
    finding: Finding,
    register: str | None = None,
    preset: str | None = None,
    customExemplars: Iterable[Exemplar] = (),
) -> dict:
    data = finding.asDict()
    exemplar = exemplarFor(finding.rule, preset, customExemplars)
    if exemplar:
        data["exemplar"] = exemplarInRegister(exemplar, register).asDict()
    return data


def renderJson(
    results: dict[str, list[Finding]],
    audits: dict[str, AuditResult] | None = None,
    configLabel: str | None = None,
    registers: dict[str, str] | None = None,
    preset: str | None = None,
    customExemplars: Iterable[Exemplar] = (),
) -> str:
    files = []
    for path, findings in results.items():
        register = registers.get(path) if registers else None
        entry: dict = {
            "path": path,
            "findings": [findingWithExemplar(f, register, preset, customExemplars) for f in findings],
        }
        if audits and path in audits:
            entry["audit"] = audits[path].asDict()
        files.append(entry)
    data: dict = {"version": 1}
    if configLabel is not None:
        data["config"] = configLabel
    data["files"] = files
    return json.dumps(data, ensure_ascii=False, indent=2)
